java/squeek/appleskin/client/HUDOverlayHandler.java

```diff
@@ -134,68 +151,157 @@ public static void drawSaturationOverlay(MatrixStack matrixStack, float saturati
+		int endSaturationBar = (int)Math.ceil(modifiedSaturation / 2.0F);
+
+		if (saturationGained != 0)
+			startSaturationBar = (int)Math.max(saturationLevel / 2.0F, 0);
+
+		int iconSize = 9;
+
+		for (int i = startSaturationBar; i < endSaturationBar; ++i)
 		{
-			int x = left - i * 8 - 9;
-			int y = top;
-			float effectiveSaturationOfBar = (saturationLevel + saturationGained) / 2 - i;
+			// gets the location that needs to be render of icon
+			Point2D location = foodBarOffsets.get(i);
+			if (location == null)
+				continue;
+
+			int x = (int)location.getX() + 0;
+			int y = (int)location.getY() + 0;
+
+			int v = 0;
+			int u = 0;
+
+			float effectiveSaturationOfBar = (modifiedSaturation / 2.0F) - i;
 
 			if (effectiveSaturationOfBar >= 1)
-				mc.inGameHud.drawTexture(matrixStack, x, y, 27, 0, 9, 9);
+				u = 3 * iconSize;
 			else if (effectiveSaturationOfBar > .5)
-				mc.inGameHud.drawTexture(matrixStack, x, y, 18, 0, 9, 9);
+				u = 2 * iconSize;
 			else if (effectiveSaturationOfBar > .25)
-				mc.inGameHud.drawTexture(matrixStack, x, y, 9, 0, 9, 9);
-			else if (effectiveSaturationOfBar > 0)
-				mc.inGameHud.drawTexture(matrixStack, x, y, 0, 0, 9, 9);
+				u = 1 * iconSize;
+
+			mc.inGameHud.drawTexture(matrixStack, x, y, u, v, iconSize, iconSize);
 		}
-		disableAlpha(alpha);
 
 		// rebind default icons
 		mc.getTextureManager().bindTexture(Screen.GUI_ICONS_TEXTURE);
+		disableAlpha(alpha);
 	}
 
 	public static void drawHungerOverlay(MatrixStack matrixStack, int hungerRestored, int foodLevel, MinecraftClient mc, int left, int top, float alpha, boolean useRottenTextures)
 	{
 		if (hungerRestored == 0)
 			return;
 
-		int startBar = foodLevel / 2;
-		int endBar = (int) Math.ceil(Math.min(20, foodLevel + hungerRestored) / 2f);
-		int barsNeeded = endBar - startBar;
+		enableAlpha(alpha);
 		mc.getTextureManager().bindTexture(Screen.GUI_ICONS_TEXTURE);
 
-		enableAlpha(alpha);
-		for (int i = startBar; i < startBar + barsNeeded; ++i)
+		int modifiedFood = Math.min(20, foodLevel + hungerRestored);
+
+		int startFoodBars = foodLevel / 2;
+		int endFoodBars = (int)Math.ceil(modifiedFood / 2.0F);
+
+		int iconStartOffset = 16;
+		int iconSize = 9;
+
+		for (int i = startFoodBars; i < endFoodBars; ++i)
 		{
-			int idx = i * 2 + 1;
-			int x = left - i * 8 - 9;
-			int y = top;
-			int icon = 16;
-			int background = 1;
+			// gets the location that needs to be render of icon
+			Point2D location = foodBarOffsets.get(i);
+			if (location == null)
+				continue;
+
+			int x = (int)location.getX() + 0;
+			int y = (int)location.getY() + 0;
+
+			// location to normal food by default
+			int v = 3 * iconSize;
+			int u = iconStartOffset + 4 * iconSize;
+			int ub = iconStartOffset + 1 * iconSize;
 
+			// relocation to rotten food
 			if (useRottenTextures)
 			{
-				icon += 36;
-				background = 13;
+				u += 4 * iconSize;
+				ub += 12 * iconSize;
 			}
 
+			// relocation to half food
+			if (modifiedFood < (i + 1) * 2)
```

Comment:
This should be
```suggestion
			if (modifiedFood < i * 2 + 1)
```

Reply:
in case 5.2:
  modifiedFood = 5.2
  endFoodBars = ceil(5.2/2) // i.e. 3
  5.2 < 2 * 2 + 1 == false // i.e. full food
in case 5.0:
  modifiedFood = 5.0
  endFoodBars = ceil(5.0/2) // i.e. 3
  5.0 < 2 * 2 + 1 == false // i.e. halt food

Reply:
`i * 2 + 1` is what's used by Minecraft in `InGameHud.renderStatusBars`:

```java
               if (z * 2 + 1 < i) {
                  this.drawTexture(matrices, ad, ae, aa + 36, 9 * af, 9, 9);
               }
```

It's also what was used previously:

https://github.com/squeek502/AppleSkin/blob/a2669bdd7bcf2340c79499a59210a9f8a8e95934/java/squeek/appleskin/client/HUDOverlayHandler.java#L163
https://github.com/squeek502/AppleSkin/blob/a2669bdd7bcf2340c79499a59210a9f8a8e95934/java/squeek/appleskin/client/HUDOverlayHandler.java#L177

The comparison is backwards, too, though, it should actually be:

```suggestion
			if (i * 2 + 1 < modifiedFood)
```

Reply:
2 * 2 + 1 >= 5.2 // false 
2 * 2 + 1 >= 5.0 // true

Reply:
In the `InGameHud.renderStatusBars `, the `ah * 2 + 1 < k` is full food case.
But we need check of half food case.
Maybe I should change the default to half food.

Reply:
Ah, I see. I think it's better to copy the vanilla code here, to make it easy to compare. So this would work too

```suggestion
			if (i * 2 + 1 == modifiedFood)
```

Reply:
food level as int, which is working, but health as float, in new case 4.9, this will still be problem, so use `>=` or change it to default half food.

2 * 2 + 1 == 5.2 // false
2 * 2 + 1 == 5.0 // true
2 * 2 + 1 == 4.9 // false, need to true

Reply:
```suggestion
if (i * 2 + 1 == (int)Math.ceil(modifiedFood))
```